### DISCUSSION TITLE
Can not run and debug tests with AVA in VSCode. Nodejs, Typescript, ESM modules, with aliases (absolute paths) working.

### DISCUSSION THREAD

**Discussion by korulis** [avajs/ava#3358] [Q&A]
This is a port of a [question](https://github.com/avajs/typescript/issues/55) from @ava/typescript forum as suggested by the author of that package - [novemberborn](https://github.com/novemberborn).

* What you're trying to do:

I am trying to create a Typescript project with Ava testing where I could debug the code while tests are running. My setup used to just work (as described [here](https://github.com/avajs/ava/blob/main/docs/recipes/debugging-with-vscode.md#debugging-with-the-debug-terminal)) when I was using `commonjs` modules, but after switching to `esm` I can not get it to work. I have spent hours and days looking for the right configuration and/or workaround for that.
This is my current [setup](https://github.com/korulis/nodejs-ts-esm-ava-vscode-debuggable/tree/master) (master branch on repo) after switching to `esm` modules. 

* What happened

Adding breakpoints and running the app `yarn start` correctly stops at breakpoints, but when running `yarn test` it does not stop at breakpoints. 

* What you expected to happen

I was expecting the execution to stop at breakpoints in `app.ts` and `test.test.ts` files

What Iv'e tried so far:
- Using `esm-module-alias` to create custom loader. This fixes the issue of alias paths in generated `.js` files, and allows debugging in vscode both with `"start": "rm -rf dist && NODE_ENV=development tsx watch ./src/app/index.ts",` and `    "build_and_start_ema": "tsc && node --loader=./dist/app/alias.js ./dist/app/index.js",` commands. It also allows running ava tests with `"test": "NODE_OPTIONS='--import=tsx' ava --config unittest.js",` ,but does not allow debuging. (on master branch)
- Using `tsconfig-paths` to fix aliased import paths in generated javascript - this does not even allow running the app after build phase. (on master branch)
- Using `ts-patch` to fix aliased import paths in generated javascript - this does not allow running tests at all. See repo [branch](https://github.com/korulis/nodejs-ts-esm-ava-vscode-debuggable/tree/master).
- Using '' this allows running the tests but not debugging them, and only after running command `"build_tsca": "tsc --project tsconfig.json && tsc-alias -p tsconfig.json",` , which means I can not run tests continuously using `yarn test --watch` . See repo [branch](https://github.com/korulis/nodejs-ts-esm-ava-vscode-debuggable/tree/with-tsc-alias).
- (new) Using simply `tsc` to compile/transpile my typescript into javascript into `dist/` folder and then runing with ava without the typescript config (`yarn build` + `yarn test`). See [commit](https://github.com/korulis/nodejs-ts-esm-ava-vscode-debuggable/commit/bb6a4a381fd54b910536d0ee77b8ded375bc3ba6) for this attempt. This does run the tests, but it still does not stop at vscode breakpoints. Why I expected the breakpoints to work in this case? After all, my breakpoints are in `.ts` files and I am executing strictly only `.js` files in `dist/` folder... - because when I do analogous thing for running the app `"build_and_start_ema": "tsc && node --loader=./dist/app/alias.js ./dist/app/index.js",` it DOES hit the breakpoints in `.ts` files even though I am executing only the `.js` files.

Overall my best attempts were attempt 1 and 4, but attempt 1 is better because it allows running tests continuously.
 
 How can I achieve being able to continously run and debug ava tests in vscode on a typescript project with ESM modules that uses aliased import paths, like it used to when I was using `commonjs` modules?

my "flattened" tsconfig in master branch:

```
> npx tsc --showConfig
{
    "compilerOptions": {
        "lib": [
            "es2022"
        ],
        "module": "nodenext",
        "target": "es2022",
        "strict": true,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "moduleResolution": "nodenext",
        "rootDir": "./src",
        "resolveJsonModule": true,
        "moduleDetection": "force",
        "isolatedModules": true,
        "verbatimModuleSyntax": true,
        "noUncheckedIndexedAccess": true,
        "noImplicitOverride": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noFallthroughCasesInSwitch": true,
        "strictNullChecks": true,
        "useUnknownInCatchVariables": true,
        "outDir": "./dist",
        "sourceMap": true,
        "baseUrl": "./",
        "paths": {
            "*": [
                "./src/*"
            ]
        },
        "allowSyntheticDefaultImports": true,
        "resolvePackageJsonExports": true,
        "resolvePackageJsonImports": true,
        "preserveConstEnums": true,
        "useDefineForClassFields": true,
        "strictFunctionTypes": true,
        "strictBindCallApply": true,
        "strictPropertyInitialization": true,
        "strictBuiltinIteratorReturn": true,
        "alwaysStrict": true
    },
    "files": [
        "./src/app/alias.ts",
        "./src/app/app.ts",
        "./src/app/app2.ts",
        "./src/app/index.ts",
        "./src/test/unit/test.test.ts",
        "./src/test/unit/test2.test.ts"
    ],
    "include": [
        "src/**/*"
    ],
    "exclude": [
        "node_modules/**/*"
    ]
}
```

How can I achieve being able to continously run and debug ava tests in vscode on a typescript project with ESM modules that uses aliased import paths, like it used to when I was using commonjs modules?

**Comment by guillaumebecard-dotcom** [comment]
Hi @korulis ! Have you succeed to solve your problem ? Going through the same issue rn

**Reply by korulis** [reply]
Hi, 
No I did not solve this. Moved to another company with slightly different tech stack, but still mightily curious how this should work.
